### PR TITLE
storage/stream: fix possible unaligned write on buffer flush request

### LINKED ISSUE #25471: disco_l475_iot1 don't write last small block
**Describe the bug**
UpdateHub uses subsys/dfu/img_util structure to interact with system flash. The current subsys/dfu/img_util over disco_l475_iot1 isn't work when try right last block with small data block.

**To Reproduce**
1. clone https://github.com/UpdateHub/zephyr/commits/topic/uhu_wifi
2. set variables at samples/net/updatehub/prj.conf
CONFIG_UPDATEHUB_WIFI_SSID="wifi network"
CONFIG_UPDATEHUB_WIFI_PSK="password"
3. west build -b disco_l475_iot1 samples/net/updatehub
4. sign image
 ~/<zephyr_home_dir>/bootloader/mcuboot/scripts/imgtool.py sign --key ~/<zephyr_home_dir>/bootloader/mcuboot/root-rsa-2048.pem --header-size 0x200 --align 8 --version 1.0.0 --slot-size 0x6C000 ~/<zephyr_home_dir>/zephyr/build/zephyr/zephyr.hex ~/<zephyr_home_dir>/zephyr/build/zephyr/zephyr-1.0.0.hex
5. west flash --hex-file build/zephyr/zephyr-1.0.0.hex
6. See error on terminal

**Expected behavior**
Flash write without errors.

**Impact**
Can't use any update that works with subsys/dfu/img_util.

**Screenshots or console output**
```
uart:~$ *** Booting Zephyr OS build v2.3.0-rc1-138-g1d06829ef97b  ***
[00:00:00.005,000] <inf> mcuboot: Starting bootloader
[00:00:00.006,000] <inf> mcuboot: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x1
[00:00:00.006,000] <inf> mcuboot: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[00:00:00.006,000] <inf> mcuboot: Boot source: primary slot
[00:00:00.013,000] <inf> mcuboot: Swap type: none


*** Booting Zephyr OS build v2.3.0-rc1-179-g489f25467752  ***
[00:00:00.540,000] <inf> main: UpdateHub sample app started
[00:00:00.549,000] <inf> main: Confirming the boot image
[00:00:03.632,000] <inf> main: WIFI Connected
[00:00:03.660,000] <inf> main: Starting UpdateHub polling mode
[00:00:04.208,000] <inf> updatehub: No update available
[00:01:04.794,000] <inf> updatehub: Probe metadata received
[00:01:11.282,000] <dbg> updatehub.install_update_cb: Flash: Address: 0x00000000, Size: 1024, Flush: 0
[00:01:11.849,000] <dbg> updatehub.install_update_cb: Flash: Address: 0x00000400, Size: 1024, Flush: 0
[00:01:12.416,000] <dbg> updatehub.install_update_cb: Flash: Address: 0x00000800, Size: 1024, Flush: 0
[00:01:12.983,000] <dbg> updatehub.install_update_cb: Flash: Address: 0x00000c00, Size: 1024, Flush: 0
[00:01:13.550,000] <dbg> updatehub.install_update_cb: Flash: Address: 0x00001000, Size: 1024, Flush: 0
[00:01:14.117,000] <dbg> updatehub.install_update_cb: Flash: Address: 0x00001400, Size: 1024, Flush: 0
...
[00:03:42.028,000] <dbg> updatehub.install_update_cb: Flash: Address: 0x00042400, Size: 1024, Flush: 0
[00:03:42.557,000] <dbg> updatehub.install_update_cb: Flash: Address: 0x00042800, Size: 92, Flush: 1
[00:03:42.569,000] <err> flash_stm32: Write range invalid. Offset: 845824, len: 92
[00:03:42.579,000] <err> STREAM_FLASH: flash_write error -22 offset=0x000ce800
RC from stream_flash_buffered_write: -22
RC from flash_area_close: -22
[00:03:42.600,000] <err> updatehub: Error to write on the flash
[00:04:43.727,000] <inf> updatehub: No update available
uart:~$ 
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain Zephyr SDK 0.11.2
 - Commit ec9a413983

**Additional context**
At my environment is 100% reproducible
This is the image used at UpdateHub-CE [zephyr-disco_l475_iot1-2.0.0.zip](https://github.com/zephyrproject-rtos/zephyr/files/4654490/zephyr-disco_l475_iot1-2.0.0.zip). Please, rename extension from zip to pkg.

CC @erwango @ABOSTM @nvlsianpu @otavio

### PR DESCRIPTION
**Pull request by nvlsianpu**
On buffer flush request it is very probably that write buffer
contains amount of data which is non write-block-size aligned.
Flash memory need to be write at minimal by write-block-size chunks.
This patch addresses mechanism which ensure such behavior by adding
missing bytes.

fixes #25471

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

### CODE REVIEW

File: subsys/storage/stream/stream_flash.c [hunk 1]
```diff
@@ -112,6 +112,8 @@ int stream_flash_buffered_write(struct stream_flash_ctx *ctx, const u8_t *data,
 	int processed = 0;
 	int rc = 0;
 	int buf_empty_bytes;
+	size_t addenum;
```
Comment (pabigot) [inline subsys/storage/stream/stream_flash.c:115]: `fill_length` would be a better name than `addendum`

File: subsys/storage/stream/stream_flash.c [hunk 2]
```diff
@@ -144,6 +146,25 @@ int stream_flash_buffered_write(struct stream_flash_ctx *ctx, const u8_t *data,
...
+			 */
+			rc = flash_read(ctx->fdev,
+					ctx->offset + ctx->bytes_written,
+					(void *)&filler,
+					1);
+
+			if (rc != 0) {
+				return rc;
+			}
+			memset(ctx->buf + ctx->buf_bytes, filler, addenum);
```
Comment (pabigot) [inline subsys/storage/stream/stream_flash.c:164]: This is at risk for writing past the end of `ctx->buf`.  That should be avoided by adding a check in `stream_flash_init()` that `buf_len` is a multiple of the write block size.

**Review by pabigot — CHANGES_REQUESTED** [review]

File: subsys/storage/stream/stream_flash.c (hunk 2 above)
Comment (nvlsianpu) [reply subsys/storage/stream/stream_flash.c:165]: agree - was thought about such check before, but thought it is overkill

File: subsys/storage/stream/stream_flash.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (hakonfam) [reply subsys/storage/stream/stream_flash.c:115]: I forgot to hit "send". @pabigot 's proposal is better.

**Comment by nandojve** [comment]
I was wondering if a regression test for this would not be good.

**Comment by nvlsianpu** [comment]
> I was wondering if a regression test for this would not be good.

Yes it is possible.

File: subsys/storage/stream/stream_flash.c [hunk 3]
```diff
@@ -189,6 +212,11 @@ int stream_flash_init(struct stream_flash_ctx *ctx, struct device *fdev,
 		return -EFAULT;
 	}
 
+	if (buf_len % flash_get_write_block_size(fdev)) {
```
Comment (de-nordic) [inline subsys/storage/stream/stream_flash.c:215]: Can we do that check at the beginning of `stream_flash_init`, it looks like it is parameter check.
The `buf_len` is also used at line 200, so if it is incorrect here, then it is also incorrect at that line.

File: subsys/storage/stream/stream_flash.c [hunk 4]
```diff
@@ -144,6 +146,27 @@ int stream_flash_buffered_write(struct stream_flash_ctx *ctx, const u8_t *data,
 	}
 
 	if (flush && ctx->buf_bytes > 0) {
+		fill_length = ctx->buf_bytes %
+				 flash_get_write_block_size(ctx->fdev);
+		if (fill_length) {
```
Comment (de-nordic) [inline subsys/storage/stream/stream_flash.c:151]: This is side note: We should consider adding some API call for returning the 'erased' byte value for the device. It is at least second time when I see that the value is needed and is either read from device (here) or predefined in code, e.g. in mcuboot.

File: subsys/storage/stream/stream_flash.c [hunk 5]
```diff
@@ -144,6 +146,27 @@ int stream_flash_buffered_write(struct stream_flash_ctx *ctx, const u8_t *data,
...
+			/*
+			 * Leverage the fact that unwritten memory
+			 * should be erased in order to get the erased
+			 * byte-value.
+			 */
+			rc = flash_read(ctx->fdev,
```
Comment (de-nordic) [inline subsys/storage/stream/stream_flash.c:157]: This is side note: We should consider adding some API call for returning the 'erased' byte value for the device. It is at least second time when I see that the value is needed and is either read from device (here) or predefined in code, e.g. in mcuboot.

**Review by de-nordic — CHANGES_REQUESTED** [review]

**Review by de-nordic — APPROVED** [review]
Notes left. otherwise I think that the solution is fine.

File: subsys/storage/stream/stream_flash.c (hunk 5 above)
Comment (nvlsianpu) [reply subsys/storage/stream/stream_flash.c:157]: have this on my API request change list

**Review by pabigot — APPROVED** [review]

**Comment by nvlsianpu** [comment]
@nandojve Test added

File: tests/subsys/storage/stream/stream_flash/testcase.yaml [hunk 6]
```diff
@@ -2,6 +2,10 @@ tests:
   storage.stream_flash:
     platform_whitelist: native_posix native_posix_64
     tags: stream_flash
+storage.stream_flash.dword_wbs:
+    extra_args: DTC_OVERLAY=./unaligned_flush.overlay
```
Comment (pabigot) [inline tests/subsys/storage/stream/stream_flash/testcase.yaml:6]: Does this work?  `DTC_OVERLAY` doesn't appear to be defined.  `DTC_OVERLAY_FILE` is.

In any case using a relative path seems unusual (contrast with `OVERLAY_CONFIG=` below).

Comment (nvlsianpu) [reply tests/subsys/storage/stream/stream_flash/testcase.yaml]: fixed

**Review by hakonfam — APPROVED** [review]
There is a typo in the last commit "steram" -> "stream"

**Review by nandojve — APPROVED** [review]
After apply changes, the board disco\_l475\_iot1 started to work properly.

**Comment by nvlsianpu** [comment]
In the last commit added note on requirements for size of the stream buffer.

**Pull request merged**